### DISCUSSION TITLE
dynamically pass input and output fields

### DISCUSSION THREAD

**Discussion by NDayakarChowdary** [dbosoft/YaNco#267] [Q&A]
dynamically specifying input and output fields through JSON data.
// Currently facing challenges with the CallFunction method and mapping structures dynamically.
// Would appreciate documentation or examples demonstrating the correct usage in such scenarios.
using Dbosoft.YaNco;
using LanguageExt.ClassInstances;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

var settingsJson = @"
{
    ""ashost"": ""SERVER"",
    ""sysnr"": ""001"",
    ""client"": ""001"",
    ""user"": ""userid"",
    ""passwd"": ""Password"",
    ""lang"": ""EN""
}";

// Deserialize JSON into a dictionary
var settings = JsonConvert.DeserializeObject<Dictionary<string, string>>(settingsJson);

var connectionBuilder = new ConnectionBuilder(settings);
var connFunc = connectionBuilder.Build();

using (var context = new RfcContext(connFunc))
{

    await context.CallFunction("BAPI_EQMT_DETAIL",
        Input: f => f
            .SetField("EQUIPMENT", "20000000")
            .SetField("FIELDNAME2", "BNAME"),
        Output: f => f
                 //.MapStructure("EQUIMASTER", s => s
                 //.GetField<string>("EQUIPMENT")
                 .MapStructure("EQUIMASTER", s =>
                from EQUIPMENT in s.GetField<string>("EQUIPMENT")
                from EQUICATGRY in s.GetField<string>("EQUICATGRY")
                from EQUITYPE in s.GetField<string>("EQUITYPE")

                select (EQUIPMENT, EQUICATGRY, EQUITYPE)

            ))
        .Match(
            r => Console.WriteLine($"Result: {JsonConvert.SerializeObject(r)}"),
            l => Console.WriteLine($"Error: {l.Message}")
        );
}

input and output would dynamically pass through Json input ,have multiple setfeilds/getfields

**Comment by fw2568** [comment]
I have updated the title to make it better readable. Original title formated:

dynamically pass input and output fields through JSON data 

``` json
"function":{
   "name":"BAPI_EQMT_DETAIL",
   "inputFields":[
      {
         "fieldName":"EQUIPMENT",
         "value":"20000000"
      },
      {
         "fieldName":"FIELDNAME2",
         "value":"BNAME"
      }
   ],
   "outputFields":[
      {
         "structureName":"EQUIMASTER",
         "fieldName":"EQUIPMENT"
      },
      {
         "structureName":"EQUIMASTER",
         "fieldName":"EQUICATGRY"
      },
      {
         "structureName":"EQUIMASTER",
         "fieldName":"EQUITYPE"
      }
   ]
}

```

**Comment by fw2568** [comment]
Hello NDayakarChowdary,

It's by design that there is no sample for dynamic json input processing. As SAP functions are fixed it doesn't make much sense to build them fully generic from the input side. 

For output it may make sense for generic logging, there is a sample here: https://github.com/dbosoft/YaNco/blob/c1070b33636f16fa53bae66c3424eb34eda3de9f/test/SAPSystemTests/Program.cs#L410

However even in this case I recommend to to explicit mapping of all fields from SAP instead of a fully dynamic call. 
I don't know your business case, but in a typical architecture you first map your json to a DTO and then you fill the function call from the DTO. Same for Output. Doing so, you have a clear and well definied stable interface for your callers. 

Best Regards,
Frank
